### PR TITLE
Implement content link re-writing

### PR DESCRIPTION
**Pull request by khorwood**
Fixes: #31

### CODE REVIEW

**Comment by khorwood** [comment]
The refactored API made this much easier to do.

File: src/index.js [hunk 1]
```diff
@@ -15,8 +16,11 @@ async function run() {
 		manifestPath: validator.getManifestPath()
 	};
 	const uploader = new UploadCourseContent(inputs, api);
+	const rewriter = new LinkRewriter(inputs.contentDirectory, api);
 
-	await uploader.uploadCourseContent(validator.getInstanceDomain(), validator.getCourseOrgUnitId());
+	const manifest = await uploader.uploadCourseContent(validator.getInstanceDomain(), validator.getCourseOrgUnitId());
+
+	await rewriter.rewriteLinks(validator.getInstanceDomain(), validator.getCourseOrgUnitId(), manifest);
```
Comment (PeteLloyd) [inline src/index.js:23]: for consideration - it looks like uploadCourseContent and rewriteLinks are both updating the content.  Are you worried about how non-atomic this is?   Someone could be using the course while this runs, and we're updating old content with links that are known to be broken, and then updating them later.  If there's an error then the whole course could be left in a state where links are broken.  I personally think it's cleaner if the upload course content step only creates new objects and builds the manifest map, and all updates are leftt o the rewrite links step.  That would be more atomic, so each existing object will be left in either the old or new state and never an in-between state.   I don't think it's much work to fix, but I'd be happy creating an issue if you don't think it's worth fixing at this point.

Comment (khorwood) [reply src/index.js:23]: I think I'd opt for an issue at this point.
Unless this workflow makes sense:
- We're comfortable creating the content with empty HTML documents on the first pass. (For file topics the file is mandatory :S)
- Update the content based on created manifest
- Update all the content with the updated HTML

File: src/link-rewriter.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		const links = this._getAllLinks(document);
+
+		let isDirty = false;
+		for (const link of links) {
+			const href = link.attrs.find(x => x.name === 'href');
+			if (!href || href.value.includes('://') || href.value.startsWith('/') || href.value.startsWith('#')) {
+				continue;
+			}
+
+			console.log(`Found relative link: '${href.value}' in '${item.descriptionFileName || item.fileName}`);
```
[the remark this answers is not in the document]
Comment (khorwood) [reply src/link-rewriter.js:42]: Just status, but can be removed.

File: src/link-rewriter.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+			const targetFileName = href.value.replace('../', '').replace(/.html$/, '.md');
+			const target = manifest.find(x => x.descriptionFileName === targetFileName || x.fileName === targetFileName);
+			if (!target) {
+				throw new Error('Could not find matching content item');
+			}
+
+			const newHref = this._getNewHref(orgUnit, item, target);
+
+			console.log(`Updating link: '${href.value}' => '${newHref}'`);
```
[the remark this answers is not in the document]
Comment (khorwood) [reply src/link-rewriter.js:52]: Just status, but can be removed.

File: src/index.js (hunk 1 above)
Comment (PeteLloyd) [reply src/index.js:23]: I think I am fine with that, but I'll create an issue anyway.  It can be addressed in a future PR.

**Pull request merged**